src/main/java/com/compstore/repository/ProductRepository.java

```diff
@@ -0,0 +1,14 @@
+package com.compstore.repository;
+
+import com.compstore.entity.ProductEntity;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ProductRepository<T extends ProductEntity> extends JpaRepository<T, UUID> {
+
+    List<T> findByIdIn(Set<UUID> uuids);
```

Comment:
Repozytorium może zwrócić obiekty różnych typów.

Tutaj zastanawiałem się nad podejściem, bo tak jak mówi artykuł który wklejałem powyżej, kiedy repozytorium zwraca obiekty różnych typów to musi wykonać się trochę joinów i to może mocno obniżać wydajność (bo konieczny jest dostęp do wielu tabel). Właściwie to można by zwrócić ProductEntity, gdyż w jedynym miejscu, gdzie ta metoda jest wykorzystywana to nie musimy mieć dostępu do pól specyficznych dla typów dziedziczących (pc, smartfon itp), tylko do pól bazowych np. cena. Więc zwrócenie samych Products mogłoby wystarczyć. Ciekawe, czy to pozwoliłoby uniknąć wspomnianych joinów.
Pomyślę o tym do momentu aż sprawdzić review.
Jeszcze zastanawiam się jakby miał wyglądać dostęp do obiektów np pc, smartfon z takiej listy produktów. Trzeba by przechodzić przez listę zwrotną i mapować obiekty do typów javovych. Tylko na jakiej podstawie? Chat mi na razie podpowiedział, że instanceof. Może to okazać sie konieczne, kiedy będziemy chcieli wyświetlać te produkty z zamówienia - no chyba, że zdecydujemy, że odwołamy się tylko do skonkatenowanego opisu, który także będzie w tej bazowej tabeli Products

Reply:
Nawet jak `findByIdIn` wykonuję dla listy ProductEntity to i tak joiny się robią.
Ogólnie posiadanie oddzielnej tabeli Products jest konieczne, gdyż w zamówieniu muszę posłużyć się kluczem obcym do produktu. Jeśli by nie było wspólnej tabeli Products to nie wiadomo z jakiej tabeli miałby pochodzić klucz obcy (bo byłoby wiele - PC, Smartfony itp)